perceval/backends/github.py

```diff
@@ -0,0 +1,250 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+#
+# Authors:
+#   Alvaro del Castillo San Felix <acs@bitergia.com>
+#
+
+
+
+import json
+import logging
+import os.path
+
+import requests
+
+from ..backend import Backend, BackendCommand
+from ..cache import Cache
+from ..errors import CacheError
+from ..utils import DEFAULT_DATETIME, str_to_datetime
+
+logger = logging.getLogger(__name__)
+
+class GitHub(Backend):
+    """GitHub backend for Perceval
+
+    This class allows the fetch the issues stored in GitHub
+    repository.
+    """
+
+    def __init__(self, owner=None, repository=None, token=None,
+                 cache=None):
+        """ Init a GitHub instance
+        :param owner: GitHub owener
+        :param repository: GitHub repository from the owner
+        :param token: GitHub auth token to access the API
+        :param cache: Use issues already retrieved in cache
+        """
+        super().__init__(cache=cache)
+        self.owner = owner
+        self.repository = repository
+        self.client = GitHubClient(owner, repository, token)
+
+    def fetch(self, from_date=DEFAULT_DATETIME):
+        """Fetch the issues from the repository.
+
+        The method retrieves, from a GitHub repository, the issues
+        updated since the given date.
+
+        :param from_date: obtain issues updated since this date
+
+        :returns: a generator of issues
+        """
+
+        self._purge_cache_queue()
+
+        issues = self.client.get_issues(from_date)
+
+        self._push_cache_queue(issues)
+        self._flush_cache_queue()
+
+        while issues:
+            issue = issues.pop(0)
+            yield issue
+
+            if not issues:
+                issues = self.client.get_issues()
+                if issues and len(issues)>0:
+                    self._push_cache_queue(issues)
+                    self._flush_cache_queue()
+
+    def fetch_from_cache(self):
+        """Fetch the issues from the cache.
+
+        It returns the issues stored in the cache object provided during
+        the initialization of the object. If this method is called but
+        no cache object was provided, the method will raise a `CacheError`
+        exception.
+
+        :returns: a generator of items
+
+        :raises CacheError: raised when an error occurs accessing the
+            cache
+        """
+        if not self.cache:
+            raise CacheError(cause="cache instance was not provided")
+
+        cache_items = self.cache.retrieve()
+
+        for issues in cache_items:
+            for issue in issues:
+                yield issue
+
+
+class GitHubClient:
+
+    def __init__(self, owner, repository, token):
```

Comment:
El mar, 22-12-2015 a las 17:06 -0800, Santiago Dueñas escribió:
> In perceval/backends/github.py:
> > +        :raises CacheError: raised when an error occurs accessing
> the
> > +            cache
> > +        """
> > +        if not self.cache:
> > +            raise CacheError(cause="cache instance was not
> provided")
> > +
> > +        cache_items = self.cache.retrieve()
> > +
> > +        for issues in cache_items:
> > +            for issue in issues:
> > +                yield issue
> > +
> > +
> > +class GitHubClient:
> > +
> > +    def __init__(self, owner, repository, token):
> The API is the same. The only thing that changes is the base URL of
> the server. For now, it is OK but it would be nice to take into
> account to develop the backend as much closer as possible to how it
> will work for both platforms.

Great. Being that the case, we could change it now, but I have no
environment to test it. I will add a new base URL param optional.

> —
> Reply to this email directly or view it on GitHub.

-- 
Alvaro del Castillo San Félix
acs@bitergia.com - Chief Technical Officer (CTO)
http://www.bitergia.com
"Software metrics for your peace of mind"